cloudinit/net/dhcp.py

```diff
@@ -149,6 +157,19 @@ def dhcp_discovery(dhclient_cmd_path, interface, dhcp_log_func=None):
         "-sf",
         "/bin/true",
     ]
+    is_ib_interface = read_sys_net_safe(interface, "type") == "32"
+    if is_ib_interface:
+        dhcp_client_identifier = "20:%s" % get_interface_mac(interface)[36:]
```

Comment:
Why not use the whole GUID? What's the point of this? ~~Mellanox~~ Nvidia's docs don't suggest that in the example I saw, so I'm curious why this approach here.

Reply:
the dhcp over ipoib spec [1] doesn't dictate what should be the client-id value.  redhat change the client-id generation in newer dhcp client. The only way it working consistant across the centos7/8/9 is using the following flag send dhcp-client-identifier = hardware; which send the client-id in format of <hw_type>+ <hw address>. 
So for ib it will be hardware-type 32 + <guid>


[1]  - https://www.rfc-editor.org/rfc/rfc4390.html
[2] - https://bugzilla.redhat.com/show_bug.cgi?id=1814386

Reply:
@esposem Apologies for the direct ping - can get access to [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=1814386)? I tried requesting group membership, but that failed.

Reply:
what is you email address I can try to add you?

Reply:
try now I remove it from being private

Reply:
> try now I remove it from being private

Excellent, thank you. Do I understand correctly that this truncated hwaddr format is required for backwards compatibility with centos7?

Reply:
so according to redhat if you don't specify "send dhcp-client-identifier = hardware" the won't guarantee how client-id is generated between major version.
the client-id specified in Nvidia documentation [1] will work for CentOS 7, but for CentOS 8 it won't work it will be generated differently.
The only why to enforce it to be compatible in all versions is using "send dhcp-client-identifier = hardware" which for ethernet means 01-<mac> and for IB means 20-<GUID>. because we don't want to enforce the "send dhcp-client-identifier = hardware" on all interfaces (some may be ethernet) we are generating client-id just for IB interface according to 20-<GUID>. 

by the way the we are not truncated GUID.  
the hardware address exposed for ipoib is 20 bytes [2]
1. 1 byte reserved
2. 4 bytes for QP
3. 16 bytes for GID

GID (Global Identifier)  contains the GUID (Global Unique Identifiers) [3] which are the last 8 bytes of the GID.

[1] - https://docs.nvidia.com/networking/pages/viewpage.action?pageId=49158775#IPoverInfiniBand(IPoIB)-DHCPClient(Optional)
[2] - https://www.rdmamojo.com/2015/02/16/ip-infiniband-ipoib-architecture/ 
[3] - chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://network.nvidia.com/related-docs/whitepapers/InfiniBandFAQ_FQ_100.pdf

Reply:
Thanks for the info.

I spent some time reading systemd-networkd and NetworkManager's implementations, as well as the relevant specs and docs. I agree that this is the right approach.

If I understand correctly, this way the dhcp server will see the same dhcp-client-identifier later in boot (when systemd/NetworkManager provision network) and should issue the same IP as it did to dhclient.

Reply:
yes that correct

Reply:
Hmmm, actually the codepath exists in systemd-networkd, but using hardware address (`ClientIdentifier=mac`) is not the default setting. By default systemd-networkd uses `ClientIdentifier=duid`, which means that different IP addresses will be assigned to  dhclient and systemd-networkd.

Think we can/should try to add systemd support too?

Reply:
So, as I saw that IPOIB is not supported in cloud-init for ubuntu at all, there is no implementation here [1]
Now as conclusion, we have 3 issues here

1. It sends a wrong dhcp-client-identifier for sysconfig implementation (redhat based) which this patch fixes it
2. When metadata service is not reachable for any reason, cloud-init calls fallback data source which:

- For redhat-based, it creates ifcfg files with ethernet type and so it failed to get IP address. (We need to provide a fix for it)
- For ubuntu-based, it creates ethernet configuration without configuring dhcp-identifier and so it used the default ClientIdentifier=duid and it failed to get IP address, (Should be included in the last fix above)

3. We don't have netplan implementation in cloud-init for ubuntu-based, so in case metadata service is reachable and it applied the network configuration from DS, it will skip IB devices because it won't match any case in rendered objects in netplan (We need to add support for this)

[1] https://github.com/canonical/cloud-init/blob/main/cloudinit/net/netplan.py

Reply:
Thanks, good point on netplan support being required. Since this PR is an improvement in current state, I don't think we need to hold this PR up any further - netplan IPoIB support can be implemented in later change.